### ISSUE TITLE
False negative for an unbound variable when unpacking in `with` statement

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#2649]
Unpacking the context manager's `__enter__` result is subject to exception swallowing. In other words, this:
```py
with Foo() as [x, y, z]:
    ...
```
is the same as:
```py
with Foo() as r:
    [x, y, z] = r
    ...
```
So in this case:
```py
from typing import Iterator
import contextlib

class Foo:
    def __enter__(self) -> list[int]:
        return [1, 2, 3, 4, 5]

    def __exit__(self, exc_type, exc_value, exc_tb) -> bool:
        return exc_type is ValueError

with Foo() as [x, y, z]:
    print("Unpacking succeeded!")

a = (x, y, z)
reveal_type(a)
```
`x`, `y` and `z` will be unbound at the `a = (x, y, z)` line. But Pyright doesn't report these as possibly unbound.

**Comment by decorator-factory** [comment]
I'm not sure how useful flagging this would be. I imagine that this will cause lots of false positives.

**Comment by erictraut** [comment]
The behavior is correct. The `__enter__` method is returning a `list[int]`. The `list` type doesn't have any length information, so any attempt to unpack a list is assumed to provide values for all unpacked values. 

If you are interested in catching errors like this, you should switch to tuples. The length information is retained for a tuple, so when it is unpacked, the type checker can validate that the correct number of items are provided.

```python
class Foo:
    def __enter__(self) -> tuple[int, int, int, int, int]:
        return (1, 2, 3, 4, 5)

    def __exit__(self, exc_type, exc_value, exc_tb) -> bool:
        return exc_type is ValueError

with Foo() as (x, y, z): # Tuple size mismatch
    print("Unpacking succeeded!")
```

**Comment by decorator-factory** [comment]
The issue wasn't related to lists, really. You could have a different target that could fail:
```py
with Foo() as (x.bar, y):  # y.bar assignment could fail
    ...
```
and if the failure is swallowed by the context manager, `x` will be unbound
